scene/main/node.cpp

```diff
@@ -1868,6 +1868,59 @@ TypedArray<Node> Node::find_children(const String &p_pattern, const String &p_ty
 	return ret;
 }
 
+bool Node::is_child_of_exposed_node(const Node *p_owner) const {
+	HashSet<const Node *> visited;
+	if (p_owner) {
+		const Node *n = p_owner->get_parent();
+		while (n) {
+			visited.insert(n);
+			n = n->data.parent;
+		}
+	}
+	const Node *common_parent = data.parent;
+	while (common_parent) {
+		if (!common_parent->get_scene_file_path().is_empty() || (p_owner && visited.has(common_parent))) {
+			break;
+		}
+		if (common_parent->is_exposed_in_owner()) {
+			return true;
+		}
+		common_parent = common_parent->data.parent;
+	}
+	return false;
+}
+
+bool Node::has_exposed_nodes(bool p_recursive) const {
```

Comment:
Can't it be more efficient by using `owned_exposed_nodes`? Same above probably.

Reply:
That was the original idea, but owned_exposed_nodes does not provide enough information for nodes to be found when adding them to the scene tree editor. Since we're going down the hierarchy we need to know if there are still exposed nodes to add.

https://github.com/godotengine/godot/pull/84018/files/74368148e0435069010c2ac04220693959b386a7#diff-632c25b2b2f35d31552f71c3dff7e0dda42f53d05d54a876be9e84998d919ce6R220-R224

Reply:
Do you think caching them in a similar way to acquire_exposed_node by climbing up the parent tree to owner would be a good improvement?

Reply:
> owned_exposed_nodes does not provide enough information for nodes to be found when adding them to the scene tree editor

Isn't every exposed node added to the `owned_exposed_nodes` set? What more information do you need for this method, other than the set is not empty? In case of recursive you can iterate the exposed set and do the same check.

btw `recursive` argument is always `true`, so you could as well remove it.

Reply:
> Isn't every exposed node added to the `owned_exposed_nodes` set? What more information do you need for this method, other than the set is not empty? In case of recursive you can iterate the exposed set and do the same check.

it is, but when adding nodes to the scene tree, in order to maintain the correct hierarchy i believe we'd have to parse the exposed nodes in their hierarchical order and not order in that set. So for example:
![image](https://github.com/user-attachments/assets/7b7f860e-c51a-4e57-99e6-fcecf616cb64)

Would result in a set of [Exposed, Deep_Exposed] in Spinner. But we wouldnt know that we need to nest the Deep_Exposed TreeItem under Exposed's TreeItem without parsing the tree of Spinner

Reply:
Yeah but `has_exposed_nodes()` only checks if exposed nodes exist. The hierarchy does not matter here. Same with `is_child_of_exposed_node()` (which can be determined using owners).

Reply:
You're right, but where it's used is traveling down the hierarchy where knowing a child exposed node exists is important. 

editor/gui/scene_tree_editor.cpp:220
```cpp
	} else if (p_node->has_exposed_nodes(true)) {
			for (int i = 0; i < p_node->get_child_count(); i++) {
				_add_nodes(p_node->get_child(i), p_parent);
			}
			return;
```  

As for is_child_of_exposed_node, I think you're totally right, ill swap that out for a get_owner  exposed nodes check.

Reply:
> is_child_of_exposed_node() (which can be determined using owners).

actually is_child_of_exposed_node would need to stay for similar reasons.

- Instance
  -   Non-exposed node
       - child1
  -  Exposed node
      - child2
      
child1   is_child_of_exposed_node would be false
child2   is_child_of_exposed_node would be true

Instance get_owned_exposed_nodes() just states if it has any or not, no info about child in question. unless you iterate the exposed nodes and check if their an ancestor